## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [1]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [51]:
# Write your code for importing the csv file here
movies = pd.read_csv('https://github.com/yashj1301/Python3-UpGrad-UMich/blob/master/Python%203.x/Upgrad/Modules/Module%203%20-%20Python%20for%20Data%20Science/Session%203%20-%20Pandas/Data/Movies.csv?raw=true')
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [3]:
# Write your code for inspection here
movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [4]:
movies.index

RangeIndex(start=0, stop=3853, step=1)

In [19]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

In [7]:
movies.describe().round(3)

num_critic_for_reviews  duration  director_facebook_likes  \
count                3852.000  3852.000                 3853.000   
mean                  163.047   109.921                  784.507   
std                   123.927    22.742                 3027.396   
min                     1.000    34.000                    0.000   
25%                    72.000    95.000                   10.000   
50%                   134.000   106.000                   58.000   
75%                   221.000   120.000                  222.000   
max                   813.000   330.000                23000.000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count                3847.000                3853.000  3.853000e+03   
mean                  747.290                7575.559  5.096547e+07   
std                  1841.624               15405.383  6.932033e+07   
min                     0.000                   0.000  1.620000e+02   
25%                   183.000                 721.000  6.830957e+06   
50%                   427.000                1000.000  2.790000e+07   
75%                   685.000               12000.000  6.550000e+07   
max                 23000.000              640000.000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count         3853.000                   3853.000              3847.000   
mean        102424.005                  11239.480                 1.378   
std         150294.225                  18922.949                 2.056   
min              5.000                      0.000                 0.000   
25%          17309.000                   1817.000                 0.000   
50%          50523.000                   3876.000                 1.000   
75%         124185.000                  15972.000                 2.000   
max        1689764.000                 656730.000                43.000   

       num_user_for_reviews        budget  title_year  actor_2_facebook_likes  \
count              3853.000  3.853000e+03    3853.000                3852.000   
mean                326.720  4.524817e+07    2003.062                1959.018   
std                 407.921  2.234208e+08      10.007                4472.171   
min                   1.000  2.180000e+02    1920.000                   0.000   
25%                 102.000  1.000000e+07    1999.000                 362.000   
50%                 203.000  2.400000e+07    2005.000                 664.000   
75%                 391.000  5.000000e+07    2010.000                 971.000   
max                5060.000  1.221550e+10    2016.000              137000.000   

       imdb_score  aspect_ratio  movie_facebook_likes  
count    3853.000      3781.000              3853.000  
mean        6.463         2.110              9089.546  
std         1.054         0.353             21277.375  
min         1.600         1.180                 0.000  
25%         5.900         1.850                 0.000  
50%         6.600         2.350               206.000  
75%         7.200         2.350             11000.000  
max         9.300        16.000            349000.000

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [8]:
movies.shape

(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [29]:
(movies.isna().sum()!=0).sum()

12

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [52]:
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes',
             'actor_3_name','duration','facenumber_in_poster','content_rating','country','movie_imdb_link','aspect_ratio','plot_keywords'],
             axis=1,inplace=True)

movies.columns.size

13

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [53]:
# Write you code here
(movies.isna().sum()/movies.columns.size).idxmax()

'language'

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [54]:
# Write your code for filling the NaN values in the 'language' column here
movies['language']=movies['language'].astype(str).replace('nan','English')
movies['language']

0       English
1       English
2       English
3       English
4       English
         ...   
3848    English
3849    English
3850    Spanish
3851    English
3852    English
Name: language, Length: 3853, dtype: object

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

In [55]:
(movies.language=='English').sum()

3674

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [56]:
# before conversion
movies[['budget','gross']].head()

budget        gross
0  237000000.0  760505847.0
1  300000000.0  309404152.0
2  245000000.0  200074175.0
3  250000000.0  448130642.0
4  263700000.0   73058679.0

In [57]:
# Write your code for unit conversion here
movies['budget']/=10**6
movies['gross']/=10**6

In [58]:
movies.rename(columns={'budget':'budget in millions','gross':'gross in millions'},inplace=True)
movies[['budget in millions','gross in millions']].head()

budget in millions  gross in millions
0               237.0         760.505847
1               300.0         309.404152
2               245.0         200.074175
3               250.0         448.130642
4               263.7          73.058679

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [61]:
# Write your code for creating the profit column here
movies['profit']=movies['gross in millions']-movies['budget in millions']


In [64]:
# Write your code for sorting the dataframe here
profit_max=movies[['movie_title','profit']].sort_values(by='profit',ascending=False)
profit_max.head(10)

movie_title      profit
0                                        Avatar   523.505847
28                               Jurassic World   502.177271
25                                      Titanic   458.672302
2704         Star Wars: Episode IV - A New Hope   449.935665
2748                 E.T. the Extra-Terrestrial   424.449459
16                                 The Avengers   403.279547
482                               The Lion King   377.783777
230   Star Wars: Episode I - The Phantom Menace   359.544677
64                              The Dark Knight   348.316061
419                            The Hunger Games   329.999255

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [67]:
profit_max.iloc[4,0]

'E.T. the Extra-Terrestrial\xa0'

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [68]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'
movies.columns

Index(['director_name', 'num_critic_for_reviews', 'gross in millions',
       'genres', 'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget in millions', 'title_year',
       'imdb_score', 'movie_facebook_likes', 'profit'],
      dtype='object')

In [142]:
IMDb_Top_250=movies[['imdb_score','num_voted_users','movie_title','language']][
    movies['num_voted_users']>25000].sort_values(by='imdb_score',ascending=False).head(250)

IMDb_Top_250.head()

imdb_score  num_voted_users  \
1795         9.3          1689764   
3016         9.2          1155770   
64           9.0          1676169   
2543         9.0           790926   
325          8.9          1215718   

                                         movie_title language  
1795                       The Shawshank Redemption   English  
3016                                  The Godfather   English  
64                                  The Dark Knight   English  
2543                         The Godfather: Part II   English  
325   The Lord of the Rings: The Return of the King   English

In [143]:
IMDb_Top_250.index=[i for i in range(250)]
IMDb_Top_250['Rank']=IMDb_Top_250.index+1
IMDb_Top_250.head()

imdb_score  num_voted_users  \
0         9.3          1689764   
1         9.2          1155770   
2         9.0          1676169   
3         9.0           790926   
4         8.9          1215718   

                                      movie_title language  Rank  
0                       The Shawshank Redemption   English     1  
1                                  The Godfather   English     2  
2                                The Dark Knight   English     3  
3                         The Godfather: Part II   English     4  
4  The Lord of the Rings: The Return of the King   English     5

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [145]:
Bucket=pd.Series({'[{} to {})'.format(i,i+0.5): IMDb_Top_250['imdb_score'].between(i,i+0.4).sum() if i!=9.5 
        else IMDb_Top_250['imdb_score'].between(i,i+0.5).sum() for i in np.arange(7.5,10,0.5) })

Bucket.idxmax()

'[8.0 to 8.5)'

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [172]:
# Write your code for creating three new dataframes here
Meryl_Streep=movies[movies['actor_1_name']=='Meryl Streep'][['actor_1_name','movie_title','num_critic_for_reviews','num_user_for_reviews']]
Meryl_Streep.head()

actor_1_name             movie_title  num_critic_for_reviews  \
392   Meryl Streep       It's Complicated                    187.0   
1038  Meryl Streep         The River Wild                     42.0   
1132  Meryl Streep          Julie & Julia                    252.0   
1322  Meryl Streep  The Devil Wears Prada                    208.0   
1390  Meryl Streep        Lions for Lambs                    227.0   

      num_user_for_reviews  
392                  214.0  
1038                  69.0  
1132                 277.0  
1322                 631.0  
1390                 298.0

In [173]:
Leo_Caprio =movies[movies['actor_1_name']=='Leonardo DiCaprio'][['actor_1_name','movie_title','num_critic_for_reviews','num_user_for_reviews']]
Leo_Caprio.head()

actor_1_name        movie_title  num_critic_for_reviews  \
25   Leonardo DiCaprio           Titanic                    315.0   
49   Leonardo DiCaprio  The Great Gatsby                    490.0   
94   Leonardo DiCaprio         Inception                    642.0   
173  Leonardo DiCaprio      The Revenant                    556.0   
246  Leonardo DiCaprio       The Aviator                    267.0   

     num_user_for_reviews  
25                 2528.0  
49                  753.0  
94                 2803.0  
173                1188.0  
246                 799.0

In [174]:
Brad_Pitt = movies[movies['actor_1_name']=='Brad Pitt'][['actor_1_name','movie_title','num_critic_for_reviews','num_user_for_reviews']]
Brad_Pitt.head()

actor_1_name                           movie_title  \
97     Brad Pitt  The Curious Case of Benjamin Button    
142    Brad Pitt                                 Troy    
243    Brad Pitt                       Ocean's Twelve    
244    Brad Pitt                     Mr. & Mrs. Smith    
367    Brad Pitt                             Spy Game    

     num_critic_for_reviews  num_user_for_reviews  
97                    362.0                 822.0  
142                   220.0                1694.0  
243                   198.0                 627.0  
244                   233.0                 798.0  
367                   142.0                 361.0

In [176]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep,Leo_Caprio,Brad_Pitt],axis=0)
Combined.head()

actor_1_name             movie_title  num_critic_for_reviews  \
392   Meryl Streep       It's Complicated                    187.0   
1038  Meryl Streep         The River Wild                     42.0   
1132  Meryl Streep          Julie & Julia                    252.0   
1322  Meryl Streep  The Devil Wears Prada                    208.0   
1390  Meryl Streep        Lions for Lambs                    227.0   

      num_user_for_reviews  
392                  214.0  
1038                  69.0  
1132                 277.0  
1322                 631.0  
1390                 298.0

In [178]:
# Write your code for grouping the combined dataframe for finding the mean of critic reviews and audience reviews here
Combined.pivot_table(index='actor_1_name',values=['num_critic_for_reviews','num_user_for_reviews'],aggfunc=np.mean)

num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [182]:
mean_user=Combined.pivot_table(index='actor_1_name',values=['num_critic_for_reviews'],aggfunc=np.mean).idxmax()
mean_user

num_critic_for_reviews    Leonardo DiCaprio
dtype: object

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [181]:
mean_critic=Combined.pivot_table(index='actor_1_name',values=['num_user_for_reviews'],aggfunc=np.mean).idxmax()
mean_critic

num_user_for_reviews    Leonardo DiCaprio
dtype: object